openprescribing/matrixstore/build/download_prescribing.py

```diff
@@ -0,0 +1,257 @@
+"""
+Download prescribing data from BigQuery to gzipped CSV files in the
+"matrixstore_import" directory inside `settings.PIPELINE_DATA_BASEDIR`
+"""
+import glob
+import logging
+import os
+
+from django.conf import settings
+
+from gcutils.bigquery import Client, StorageClient
+
+from .common import (
+    get_prescribing_filename, get_filename_for_download, get_temp_filename
+)
+from .dates import generate_dates
+from .sort_and_merge_csv_files import sort_and_merge_csv_files
+
+
+logger = logging.getLogger(__name__)
+
+
+# We need a way to indicate that an export from a BigQuery table has completed
+# (sadly BigQuery doesn't do this for us) so we create a "sentinal" file once
+# the export has finished using the suffix below
+SENTINAL_SUFFIX = 'done'
+
+
+def download_prescribing(end_date, months=None):
+    # Getting a local copy of prescribing data for a given month is a
+    # multi-stage process:
+    #
+    # 1. Extract the data for that month from the main prescribing table into
+    #    its own table
+    # 2. Export that table to Google Cloud Storage (which will end up sharded
+    #    into multiple files)
+    # 3. Download those shard files
+    # 4. Consolidate the shards into a single file, sorted by BNF code
+    bq_client = Client('prescribing_export')
```

Comment:
From reading through the docs on this I don't think it's going to help us much. We still need the sentinel file as that what tells us that the export from BigQuery has finished successfully. And we still need to sort the file after downloading. Given that we aggregate the shards as part of the sort, it doesn't save us much code to have a single downloaded file. Plus the file composition will just blindly cat the CSV files together, headers and all. Whereas we want the headers only from the first file.

I also don't think the component-count and integrity checking would be much of a help. If we think there are 12 shard files and we stick them all in a composite object it will confirm that the composite object does indeed contain 12 files. But it won't tell us that there were supposed to be 13 shard files and we made an error in our pattern matching or something.

From what I can tell, this feature is mostly helpful when you want to _upload_ a large amount of data in small parts and then create a single aggregate object from it. I don't think it buys you much for downloading.